pkg/engine/engine.go

```diff
@@ -280,6 +280,11 @@ func (e *engine) invokeRuleHandler(
 					s := stringutils.JoinNonEmpty([]string{"preconditions not met", msg}, "; ")
 					return resource, handlers.WithSkip(rule, ruleType, s)
 				}
+				// substitute properties
+				if err := internal.SubstitutePropertiesInRule(logger, &rule, policyContext.JSONContext()); err != nil {
+					logger.Error(err, "failed to substitute variables in rule properties")
+					return resource, handlers.WithError(rule, ruleType, "failed to substitute variables in rule properties", err)
+				}
```

Comment:
I am adding the logic for variable substitution for properties in the ruleInvokeHandler itself because it looks like a common logic. This modifies the rule variable.